TELCODOCS-469 - Document GitOps ZTP specifics of hub-side CR templating (#53209)
openshift/openshift-docs: modules/ztp-syncing-new-configmap-changes-to-existing-pgt-crs.adoc

:_content-type: PROCEDURE
[id="ztp-syncing-new-configmap-changes-to-existing-pgt-crs_{context}"]
= Syncing new ConfigMap changes to existing PolicyGenTemplate CRs

.Prerequisites

* You have installed the OpenShift CLI (`oc`).

* You have logged in to the hub cluster as a user with `cluster-admin` privileges.

* You have created a `PolicyGenTemplate` CR that pulls information from a ConfigMap CR using hub templating.

.Procedure

. Update the contents of your `ConfigMap` CR, and apply the changes in the hub cluster.

. To sync the contents of the updated `ConfigMap` CR to the deployed policy, do either of the following:

.. Optional: Delete the existing policy and recreate it using the `PolicyGenTemplate` CR.

... Delete the policy, for example:
+
[source,terminal]
----
$ oc delete policy <policy-name> -n <policy-namespace>
----
+
... Re-apply the `PolicyGenTemplate` CR:
+
[source,terminal]
----
$ oc apply -f sriov-pgt-example.yaml
----

.. Optional: Apply a special annotation `policy.open-cluster-management.io/trigger-update` to the policy with a different value every time when updating the `ConfigMap`. For example, run the following command:
+
[source,terminal]
----

> $ oc label <policy-name> -n <policy-namespace> policy.open-cluster-management.io=trigger-update

Reviewer: ```suggestion
$ oc label <policy-name> -n <policy-namespace> policy.open-cluster-management.io/trigger-update='1'
```

----

> ... Re-apply the `PolicyGenTemplate` CR to regenerate the `ClusterGroupUpdate` CR. For example, run the following command:

Reviewer: ```suggestion
... Re-apply the `PolicyGenTemplate` CR to update the policy. For example, run the following command:
```

+
[source,terminal]
----
$ oc delete clustergroupupgrade <cgu-name> -n <cgu-namespace>
----

... Re-apply the `PolicyGenTemplate` CR to regenerate the `ClusterGroupUpdate` CR. For example, run the following command:
+
[source,terminal]
----

> $ oc apply -f sriov-pgt-example.yaml

Reviewer: This doesn't look right to me. After the label is updated in PGT, should the next step be "oc apply -f sriov-pgt-example.yaml" to update the policy? then create a new CGU to rollout the change made in configmap

> ... If it exists, delete the existing `ClusterGroupUpdate` CR that contains the policy. Create a new `ClusterGroupUpdate` CR to apply the updated `ConfigMap` changes. For example:
> +
> [source,terminal]
> ----
> $ oc delete clustergroupupgrade <cgu_name> -n <cgu_namespace>
> ----

Reviewer: This should be step3.
```suggestion
.  Delete the existing `ClusterGroupUpdate` CR that contains the policy if it exists (optional). Create a new `ClusterGroupUpdate` CR with the policy included to apply the updated `ConfigMap` changes. For example:
+
[source,terminal]
----
$ oc delete clustergroupupgrade <cgu_name> -n <cgu_namespace>
$ oc apply -f clustergroupupgrade.yaml
----
```

Author: Updated. Could you give me an representative example CGR (`clustergroupupgrade.yaml`) for this please @Missxiaoguo.

Author: @Missxiaoguo I added an example CGU CR. PTAL.

Author: Question: do we need to include the list of clusters in the updated CGU CR?
```
spec:
  managedPolicies:
    - <managed_policy>
  enable: false
  clusters: <1>
  - <managed_cluster_1>
  - <managed_cluster_2>
```

Author: @Missxiaoguo does this look OK?

Reviewer: Set `enable: true`, others look good.

Author: updated.

// Module included in the following assemblies:
//
// * scalability_and_performance/ztp_far_edge/ztp-advanced-policy-config.adoc

:_content-type: PROCEDURE
[id="ztp-syncing-new-configmap-changes-to-existing-pgt-crs_{context}"]
= Syncing new ConfigMap changes to existing PolicyGenTemplate CRs

.Prerequisites

* You have installed the OpenShift CLI (`oc`).

* You have logged in to the hub cluster as a user with `cluster-admin` privileges.

* You have created a `PolicyGenTemplate` CR that pulls information from a `ConfigMap` CR using hub cluster templates.

.Procedure

. Update the contents of your `ConfigMap` CR, and apply the changes in the hub cluster.

. To sync the contents of the updated `ConfigMap` CR to the deployed policy, do either of the following:

> .. Optional: Delete the existing policy and recreate it using the `PolicyGenTemplate` CR.

Reviewer: ```suggestion
..Option 1: Delete the existing policy and the policy will be recreated by ArgoCD immediately with the `PolicyGenTemplate` CR.
```

... Delete the policy, for example:
+
[source,terminal]
----
$ oc delete policy <policy_name> -n <policy_namespace>
----
+
... Re-apply the `PolicyGenTemplate` CR:
+
[source,terminal]
----

> $ oc apply -f sriov-pgt-example.yaml

Reviewer: Should be removed. ArgoCD will automatically recreate the policy when the policy is deleted by the above delete command.

----

> .. Optional: Apply a special annotation `policy.open-cluster-management.io/trigger-update` to the policy with a different value every time when updating the `ConfigMap`. For example, run the following command:

Reviewer: ```suggestion
.. Option 2: Apply a special annotation `policy.open-cluster-management.io/trigger-update` to the policy with a different value every time when updating the `ConfigMap`. For example, run the following command:
```

+
[source,terminal]
----

> $ $ oc label <policy_name> -n <policy_namespace> policy.open-cluster-management.io/trigger-update="1"

Reviewer: This should be annotation, not label. 
```suggestion
$ $ oc annotate policy <policy_name> -n <policy_namespace> policy.open-cluster-management.io/trigger-update="1"
```

----

> ... Re-apply the `PolicyGenTemplate` CR to update the applied policy. For example, run the following command:
> +
> [source,terminal]
> ----
> $ oc apply -f sriov-pgt-example.yaml

Reviewer: This should be removed as the policy has already been updated by the above command 
```suggestion
```

----

. Optional: if it exists, delete the `ClusterGroupUpdate` CR that contains the policy. For example:
+
[source,terminal]
----
$ oc delete clustergroupupgrade <cgu_name> -n <cgu_namespace>
----

.. Create a new `ClusterGroupUpdate` CR that includes the policy to apply with the updated `ConfigMap` changes. For example add the following YAML to the file `cgr-example.yaml`:
+
[source,yaml]
----
apiVersion: ran.openshift.io/v1alpha1
kind: ClusterGroupUpgrade
metadata:
  name: <cgr_name>
  namespace: <policy_namespace>
spec:
  managedPolicies:
    - <managed_policy>
  enable: false
  clusters: <1>
  - <managed_cluster_1>
  - <managed_cluster_2>
  remediationStrategy:
    maxConcurrency: 2
    timeout: 240
----
<1> The list of clusters to update.
+
.. Apply the updated policy:
+
[source,terminal]
----
$ oc apply -f cgr-example.yaml

> ----

Reviewer: Probably add this as an additional resource at the end to indicate this limitation in ACM.
https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.6/html-single/governance/index#special-annotation-processing

Author: Added as a note inline
